process/smartContract/processProxy/processProxy.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/ElrondNetwork/elrond-go/process/smartContract/scrCommon"
+	"github.com/ElrondNetwork/elrond-go/state"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+var log = logger.GetOrCreate("processProxy")
+
+var _ scrCommon.TestSmartContractProcessor = (*scProcessorProxy)(nil)
+
+type configuredProcessor string
+
+const (
+	procV1 configuredProcessor = "processorV1"
+	procV2 configuredProcessor = "processorV2"
+)
+
+type scProcessorProxy struct {
+	configuredProcessor configuredProcessor
+	args                scrCommon.ArgsNewSmartContractProcessor
+	processor           process.SmartContractProcessorFull
+	processorsCache     map[configuredProcessor]process.SmartContractProcessorFull
+	testScProcessor     scrCommon.TestSmartContractProcessor
+	testProcessorsCache map[configuredProcessor]scrCommon.TestSmartContractProcessor
+	mutRc               sync.Mutex
+}
+
+// NewSmartContractProcessorProxy creates a smart contract processor proxy
+func NewSmartContractProcessorProxy(args scrCommon.ArgsNewSmartContractProcessor, epochNotifier vmcommon.EpochNotifier) (*scProcessorProxy, error) {
+	scProcessorProxy := &scProcessorProxy{
+		args: scrCommon.ArgsNewSmartContractProcessor{
+			VmContainer:         args.VmContainer,
+			ArgsParser:          args.ArgsParser,
+			Hasher:              args.Hasher,
+			Marshalizer:         args.Marshalizer,
+			AccountsDB:          args.AccountsDB,
+			BlockChainHook:      args.BlockChainHook,
+			BuiltInFunctions:    args.BuiltInFunctions,
+			PubkeyConv:          args.PubkeyConv,
+			ShardCoordinator:    args.ShardCoordinator,
+			ScrForwarder:        args.ScrForwarder,
+			TxFeeHandler:        args.TxFeeHandler,
+			EconomicsFee:        args.EconomicsFee,
+			TxTypeHandler:       args.TxTypeHandler,
+			GasHandler:          args.GasHandler,
+			GasSchedule:         args.GasSchedule,
+			TxLogsProcessor:     args.TxLogsProcessor,
+			BadTxForwarder:      args.BadTxForwarder,
+			EnableEpochsHandler: args.EnableEpochsHandler,
+			EnableEpochs:        args.EnableEpochs,
+			VMOutputCacher:      args.VMOutputCacher,
+			ArwenChangeLocker:   args.ArwenChangeLocker,
+			IsGenesisProcessing: args.IsGenesisProcessing,
+		},
+	}
+
+	scProcessorProxy.processorsCache = make(map[configuredProcessor]process.SmartContractProcessorFull)
+	scProcessorProxy.testProcessorsCache = make(map[configuredProcessor]scrCommon.TestSmartContractProcessor)
+
+	var err error
+	err = scProcessorProxy.createProcessorV1()
+	if err != nil {
+		return nil, err
+	}
+
+	err = scProcessorProxy.createProcessorV2()
+	if err != nil {
+		return nil, err
+	}
+
+	epochNotifier.RegisterNotifyHandler(scProcessorProxy)
+
+	return scProcessorProxy, nil
+}
+
+func (proxy *scProcessorProxy) createProcessorV1() error {
+	processor, err := smartContract.NewSmartContractProcessor(proxy.args)
+	proxy.processorsCache[procV1] = processor
+	proxy.testProcessorsCache[procV1] = smartContract.NewTestScProcessor(processor)
+	return err
```

Comment:
It looks prettier not to treat `err`, but if the constructor for `NewTestScProcessor` will add some logic using the processor, the code could panic

Reply:
fixed